src/storage/src/client/read_object.rs

```diff
@@ -413,11 +416,14 @@ impl ReadObjectResponse {
 
     #[cfg(feature = "unstable-stream")]
     #[cfg_attr(docsrs, doc(cfg(feature = "unstable-stream")))]
-    /// Convert the response to a [futures::Stream].
-    pub fn into_stream(self) -> impl futures::Stream<Item = Result<bytes::Bytes>> {
-        // TODO(#2049): implement checksum checks for streams.
+    /// Convert the response to a [Stream].
+    pub fn into_stream(self) -> impl Stream<Item = Result<bytes::Bytes>> {
         use futures::TryStreamExt;
-        self.inner.bytes_stream().map_err(Error::io)
+        let response_crc32c = self.check_crc32c();
+        Crc32CStream::new(
```

Comment:
I am not sure if this idea is any better:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=a520b10536b79869c6b70e7737a2923b

Reply:
I'm looking at your two suggestions, but would likely swap to these in a follow up PR.